# [\#7865 Issue](https://github.com/ethereum/solidity/issues/7865) `closed`: Require explicit typing for REVERT data, update require strings

#### <img src="https://avatars.githubusercontent.com/u/382183?u=cc7b2e76c56456ff05e23fa5ca044e4a461b2eb1&v=4" width="50">[fulldecent](https://github.com/fulldecent) opened issue at [2019-12-02 17:03](https://github.com/ethereum/solidity/issues/7865):

# Motivation

- Try/catch is implemented here https://github.com/ethereum/solidity/pull/7328/files
- This hardcodes in that REVERT data should always be a string
- Strings are bad, and are not required from https://eips.ethereum.org/EIPS/eip-140
- If 0.6.0 is released as-is it will be very hard to change later

# Here is how strings hurt our try/catch

It is wasteful because string comparison is done to check a status code.

```solidity
pragma solidity ^0.5;

contract Accumulator {
    uint public count;
    string public constant OVERFLOW = "Number is too high";
    
    function increment() external {
        count++;
        require(count < 10, OVERFLOW);
    }
}

contract UseAccumulator {
    Accumulator public myAccumulator;
    
    constructor(Accumulator newAccumulator) public {
        myAccumulator = newAccumulator;
    }
    
    // This function does not care if the increment fails due to overflow
    function maybeDelegateIncrement() external {
        myAccumulator.increment();
        
        try myAccumulator.increment() {
        } catch Error(string memory reason) {
            if (keccak256(Accumulator.OVERFLOW) == keccak256(reason)) {
                // ignore this error
            } else {
                revert;
            }
        } catch (bytes memory /*lowLevelData*/) {
            // This is another unexpected error
            revert;
        }
    }
}
```

# Here is how we should do it instead

This is efficient because it allows to directly compare single words rather than strings. Also it exposes errors through the ABI, which can be localized on the client.

Somebody should make a common convention and use it to return `uint` for most cases. That allows the client to make sense of most reverts without requiring access to the contract ABI.

```solidity
pragma solidity ^0.5;

contract Accumulator {
    enum Error {
        OVERFLOW
    }
    
    uint public count;

    function increment() external throws (Error) {
        count++;
        require(count < 10, OVERFLOW);
    }
}

contract UseAccumulator {
    Accumulator public myAccumulator;
    
    constructor(Accumulator newAccumulator) public {
        myAccumulator = newAccumulator;
    }
    
    // This function does not care if the increment fails due to overflow
    function maybeDelegateIncrement() external {
        myAccumulator.increment();
        
        try myAccumulator.increment() {
        } catch (Error error) {
            if (error == Accumulator.Error.OVERFLOW) {
                // ignore this error
            } else {
                revert;
            }
        } catch { // This catch clause could be excluded entirely because the unhandled exception would throw anyway
            revert;
            // This is any other type of error that cannot be properly decoded with the specified throws(Error) signature
        }
    }
}        
```

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2019-12-02 17:03](https://github.com/ethereum/solidity/issues/7865#issuecomment-561093968):

Typed errors are a planned feature. For now, the string should only always be explanatory and not be used by the program to change the logic depending on the string. Revert strings currently use a 4 byte "function selector" of "Error(string memory)". It is planned to catch exceptions based on that signature:

```
pragma solidity ^0.5;

contract Accumulator {
    error Overflow();

    uint public count;

    function increment() external {
        count++;
        require(count < 10, Overflow());
    }
}

contract UseAccumulator {
    Accumulator public myAccumulator;
    
    constructor(Accumulator newAccumulator) public {
        myAccumulator = newAccumulator;
    }
    
    // This function does not care if the increment fails due to overflow
    function maybeDelegateIncrement() external {
        myAccumulator.increment();
        
        try myAccumulator.increment() {
        } catch Accumulator.Overflow() {
                // ignore this error
        }
       // all other errors are automatically re-thrown
    }
}
```

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2019-12-02 17:03](https://github.com/ethereum/solidity/issues/7865#issuecomment-561096181):

Created an issue but maybe it already exists: https://github.com/ethereum/solidity/issues/7877


-------------------------------------------------------------------------------



[Export of Github issue for [ethereum/solidity](https://github.com/ethereum/solidity). Generated on 2022.05.23 at 03:51:38.]
